Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,535 @@
+        let channel: Channel
+
+        /// Wether the connection is currently leased or not
+        var isLeased: Bool = false
+
+        /// Indicates that this connection is about to close
+        var isClosing: Bool = false
+
+        /// Convenience property indicating wether the underlying `Channel` is active or not
+        var isActiveEstimation: Bool {
+            return self.channel.isActive
+        }
+    }
+
+    /// A connection provider of `HTTP/1.1` connections to a given `Key` (host, scheme, port)
+    ///
+    /// On top of enabling connection reuse this provider it also facilitates the creation
+    /// of concurrent requests as it has built-in politness regarding the maximum number
+    /// of concurrent requests to the server.
+    class HTTP1ConnectionProvider {
+        /// The default `EventLoop` for this provider
+        ///
+        /// The default event loop is used to create futures and is used
+        /// when creating `Channel`s for requests for which the
+        /// `EventLoopPreference` is set to `.indifferent`
+        let eventLoop: EventLoop
+
+        /// The client configuration used to bootstrap new requests
+        private let configuration: HTTPClient.Configuration
+
+        /// The key associated with this provider
+        private let key: ConnectionPool.Key
+
+        /// The `State` of this provider
+        ///
+        /// This property holds data structures representing the current state of the provider
+        /// - Warning: This type isn't thread safe and should be accessed with proper
+        /// synchronization (see the `stateLock` property)
+        fileprivate var state: State
+
+        /// The lock used to access and modify the `state` property
+        fileprivate let stateLock = Lock()
+
+        /// The maximum number of concurrent connections to a given (host, scheme, port)
+        private let maximumConcurrentConnections: Int = 8
+
+        /// The pool this provider belongs to
+        private let parentPool: ConnectionPool
+
+        /// Creates a new `HTTP1ConnectionProvider`
+        ///
+        /// - parameters:
+        ///     - key: The `Key` (host, scheme, port) this provider is associated to
+        ///     - configuration: The client configuration used globally by all requests
+        ///     - initialConnection: The initial connection the pool initializes this provider with
+        ///     - parentPool: The pool this provider belongs to
+        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, parentPool: ConnectionPool) {
+            self.eventLoop = eventLoop
+            self.configuration = configuration
+            self.key = key
+            self.parentPool = parentPool
+            self.state = State(eventLoop: eventLoop, parentPool: parentPool, key: key)
+        }
+
+        deinit {
+            self.stateLock.withLock {
```

Comment:
you can remove the locking in `deinit`, it's not necessary and may cause trouble because you don't control the thread or the stack frame `deinit` is called from.

Reply:
Done